File: examples/word-text-extract.ts
/**
 * 例子：用 Story-11.1 `Paragraph.text` 访问器把一份 docx 的所有段落抽成纯文本。
 *
 * 跑法：
 *   bun run examples/word-text-extract.ts <input.docx>
 */

import { Paragraph, WordprocessingDocument } from "../src/word/index.js";

async function main(): Promise<void> {
  const [inputPath] = process.argv.slice(2);
  if (inputPath === undefined) {
    process.stderr.write("usage: word-text-extract <input.docx>\n");
    process.exit(2);
  }

  const doc = await WordprocessingDocument.openAsync(inputPath);
  const main = doc.mainDocumentPart;
  if (main === undefined) {
    process.stderr.write("no main document part\n");
    process.exit(3);
  }

  let count = 0;
  for (const p of main.document.descendants(Paragraph)) {
    process.stdout.write(`${count.toString().padStart(4, " ")} | ${p.text}\n`);
    count += 1;
  }
  process.stderr.write(`\nextracted ${count} paragraphs\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordTextExtract.cs
// Replica of examples/word-text-extract.ts
// Opens a .docx and prints all paragraphs to stdout in the format:
//    NNN | <paragraph-text>
// where NNN is a 4-character right-padded index.
//
// Usage: called by the cross-SDK runner with a pre-generated input .docx

using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordTextExtract
{
    public static void Run(string inputPath, System.IO.TextWriter? output = null)
    {
        output ??= Console.Out;

        using var doc = WordprocessingDocument.Open(inputPath, false);
        var body = doc.MainDocumentPart?.Document?.Body;
        if (body == null) throw new InvalidOperationException("No body in document");

        int count = 0;
        foreach (var para in body.Descendants<Paragraph>())
        {
            // Collect all text from all runs (same as TS `p.text` which joins all run texts)
            var sb = new System.Text.StringBuilder();
            foreach (var run in para.Descendants<Run>())
            {
                foreach (var t in run.Descendants<Text>())
                    sb.Append(t.Text);
            }
            var text = sb.ToString();
            output.Write($"{count.ToString().PadLeft(4, ' ')} | {text}\n");
            count++;
        }
    }
}
